### PR TITLE
feat(BarChart): Added rowsConcealable prop that activates show/hide controls for barchart rows

### PR DESCRIPTION
**Pull request by t-sauer**
Deployed here: https://5afaf2fcdd6a542479baef3f--naughty-pasteur-6d413a.netlify.com/docs/packages/axiom-charts/bar-chart
You have to activate `rowsConcealable `

### CODE REVIEW

File: packages/axiom-charts/src/BarChart/BarChart.css [hunk 1]
```diff
@@ -50,3 +50,24 @@
 .ax-bar-chart__bar-label--hidden {
   opacity: 0;
 }
+
+.ax-bar-chart__conceal-button {
+  opacity: 0;
+  transition: opacity var(--transition-time-base) var(--transition-function);
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.css]: As discussed, you can shortcut all of this css with the `cloak` prop and `textColor` prop

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 2]
```diff
@@ -89,6 +96,7 @@ export default class BarChart extends Component {
   constructor(props) {
     super(props);
     this.state = {
+      hiddenRows: [],
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: How should this component deal with clearing this when the data changes? Working it out might get a bit tricky depending on how this component is getting used (for example a simple `this.props.data !== nextProps.data` check will always be `true` if a new array is always being give).

I'm wondering if the state for this should be held by the component that is rendering the BarChart, and an API like `onToggleRowVisibility()` needs to be used instead?

File: packages/axiom-components/src/Icon/IconButton.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ const sizeMap = {
 
 export default class IconButton extends Component {
   static propTypes = {
+    className: PropTypes.string,
```
Comment (HHogg) [inline packages/axiom-components/src/Icon/IconButton.js]: All of these className props should be avoidable with using the `cloak` props.

File: packages/axiom-components/src/Button/Button.js [hunk 4]
```diff
@@ -12,6 +12,7 @@ export default class Button extends Component {
     active: PropTypes.bool,
     /** Content inserted into the Button */
     children: PropTypes.node.isRequired,
+    className: PropTypes.string,
```
Comment (HHogg) [inline packages/axiom-components/src/Button/Button.js]: All of these className props should be avoidable with using the `cloak` props.

File: packages/axiom-charts/src/ChartTable/ChartTableRow.js [hunk 5]
```diff
@@ -6,12 +6,13 @@ import { Base } from '@brandwatch/axiom-components';
 export default class ChartTableRow extends Component {
   static propTypes = {
     children: PropTypes.node.isRequired,
+    className: PropTypes.string,
```
Comment (HHogg) [inline packages/axiom-charts/src/ChartTable/ChartTableRow.js]: All of these className props should be avoidable with using the `cloak` props.

File: packages/axiom-charts/src/BarChart/BarChartBars.js [hunk 6]
```diff
@@ -116,15 +119,15 @@ export default class BarChartBars extends Component {
                       size={ size } />
                 </ChartContext>
 
-                <div className={ labelClasses } style={ labelStyle }>
+                { !isConcealed && <div className={ labelClasses } style={ labelStyle }>
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChartBars.js]: If you change the `<div className="ax-bar-chart__bar-container" key={ color }>` to a Base component, you can set `cloak` once for the whole container.

File: packages/axiom-charts/src/BarChart/BarChart.js (hunk 2 above)
Comment (t-sauer) [reply packages/axiom-charts/src/BarChart/BarChart.js]: I think I initially talked about going that route with @siggysamson but I thought the decision was to keep it internally?

Comment (HHogg) [reply packages/axiom-charts/src/BarChart/BarChart.js]: As the data changes though this will need to be cleared? 

Working out when `data` is actually new data or just a new array being created at render will get tricky/impossible, so I think keeping the state for this internally would cause it to continuous clearing on a render.

**Comment by t-sauer** [comment]
I changed this PR so that the handling of which rows are hidden are handled by the consumer of the BarChart. I also changed all custom class names and used the cloak wherever possible.

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 7]
```diff
@@ -195,14 +204,24 @@ export default class BarChart extends Component {
...
+                }) }
                 hover={ index === selectedIndex }
                 key={ index }>
               <ChartTableLabel
                   textStrong={ index === selectedIndex }
                   width={ labelColumnWidth }>
-                { label }
+                <Grid horizontalGutters="tiny" verticalAlign="middle" verticalGutters={ false }>
+                  <GridCell className="ax-bar-chart__row-label">
+                    <Text onClick={ () => onToggleRowVisibility && onToggleRowVisibility(data[index]) } textColor={ onToggleRowVisibility && hidden && 'disabled' }>{ label }</Text>
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: Can the `onClick` be added to the Grid so it's only defined once?

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 8]
```diff
@@ -195,14 +204,24 @@ export default class BarChart extends Component {
...
+                  </GridCell>
+                  { onToggleRowVisibility && <GridCell shrink>
+                    <IconButton className="ax-bar-chart__conceal-button" name="preview" onClick={ () => onToggleRowVisibility(data[index]) } size="small" style="quaternary" />
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: Same here the className on IconButton won't be being applied.

Comment (HHogg) [reply packages/axiom-charts/src/BarChart/BarChart.js]: Is IconButton the correct component to use here? It looks like from the designs that it's a Icon wrapped in a Link?

File: packages/axiom-charts/src/BarChart/BarChart.css [hunk 9]
```diff
@@ -65,6 +69,28 @@
...
+.ax-bar-chart__bar-container:not(.ax-cloak--visible) .ax-bar-chart__combined-bar-diff,
+.ax-bar-chart__bar-container:not(.ax-cloak--visible) .ax-bar-chart__bar-label {
+  transition: left 0s var(--transition-function) var(--transition-time-base);
+}
+
 .ax-bar-chart__bar-label--hidden {
   opacity: 0;
 }
+
+.ax-bar-chart__conceal-button {
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.css]: This class isn't being applied anywhere. So can it be removed?

Comment (HHogg) [reply packages/axiom-charts/src/BarChart/BarChart.css]: Same with the other blocks of CSS below.

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 10]
```diff
@@ -195,14 +204,24 @@ export default class BarChart extends Component {
             xAxisLabels={ finalXAxisLabels }
             zoom={ zoom }
             zoomTo={ zoom ? zoomTo : undefined }>
-          { formattedData.map(({ values, label, benchmark }, index) =>
+          { formattedData.map(({ values, label, benchmark, hidden }, index) =>
             <ChartTableRow
+                className={ classnames('ax-bar-chart__table-row', {
+                  'ax-bar-chart__table-row--concealed': onToggleRowVisibility && hidden,
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: It doesn't look like this is needed as `ax-bar-chart__table-row--concealed` is being used to target `ax-bar-chart__conceal-button` which isn't being applied.

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 11]
```diff
@@ -195,14 +204,24 @@ export default class BarChart extends Component {
...
+                }) }
                 hover={ index === selectedIndex }
                 key={ index }>
               <ChartTableLabel
                   textStrong={ index === selectedIndex }
                   width={ labelColumnWidth }>
-                { label }
+                <Grid horizontalGutters="tiny" verticalAlign="middle" verticalGutters={ false }>
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: You shouldn't need `horizontalGutters`  or `verticalGutters={ false }` ... vertical gutters only gets applied when sibling cells wrap onto other lines, which we don't want here.

`gutters="tiny" responsive={ false }` Would be better here.

File: site/components/Documentation/Resources/Charts/BarChart.js [hunk 12]
```diff
@@ -8,6 +8,11 @@ import {
 import { dotPlotKey, dotPlotData } from './chartData';
 import DropdownContext from './DropdownContext';
 
+function onToggleRowVisibility(setValue, getValue, changedRow) {
```
Comment (stefanbuck) [inline site/components/Documentation/Resources/Charts/BarChart.js]: @HHogg ThomasS is off sick this week and I'll continue working on this. I just integrated his latest changes in Analytics and noticed that he relays on `getValue` and `changedRow` which are injected just one the documentation page. 

Maybe I miss something how this should work, but to me it looks like I have to implement this `onToggleRowVisibility` to make it working.

File: packages/axiom-charts/src/BarChart/BarChart.css (hunk 1 above)
Comment (stefanbuck) [reply packages/axiom-charts/src/BarChart/BarChart.css]: I address your comment in a8b4e240, which introduced two other issues

1. Using `textColor` prevents the hover color change so it stays grey. 


2. In the documentation `cloak` is using a css transition to toggle the visibly. For some reason I don't see this transition.

File: packages/axiom-charts/src/ChartTable/ChartTableRow.js (hunk 5 above)
Comment (stefanbuck) [reply packages/axiom-charts/src/ChartTable/ChartTableRow.js]: I revert those changes with af59e987

File: packages/axiom-charts/src/BarChart/BarChart.css (hunk 1 above)
Comment (stefanbuck) [reply packages/axiom-charts/src/BarChart/BarChart.css]: My first guess is that React does a re-render because the prop `isConcealed` changes his value.

Comment (stefanbuck) [reply packages/axiom-charts/src/BarChart/BarChart.css]: Maybe we can't use `cloak` here which is the reason why @t-sauer didn't replaced it there. His last comment says 

>  I also changed all custom class names and used the cloak wherever possible.

Comment (stefanbuck) [reply packages/axiom-charts/src/BarChart/BarChart.css]: If this is true, the only issue left is how to make this working in Analytics. See my other comment https://github.com/BrandwatchLtd/axiom/pull/559#discussion_r190529038

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 13]
```diff
@@ -195,45 +205,60 @@ export default class BarChart extends Component {
...
-                    upper={ finalUpper }
-                    values={ values } />
-              </ChartTableVisual>
-            </ChartTableRow>
-          ) }
+          { formattedData.map(({ values, label, benchmark, hidden }, index) => {
+            const Row = hidden ? 'div' : CloakContainer;
+
+            return (<Row key={ index }>
+              <ChartTableRow hover={ index === selectedIndex }>
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: The `const Row = hidden ? 'div' : CloakContainer;` isn't needed you can just give `ChartTableRow` the `cloakContainer` prop.

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 14]
```diff
@@ -195,45 +205,60 @@ export default class BarChart extends Component {
...
+                <ChartTableLabel
+                    textStrong={ index === selectedIndex }
+                    width={ labelColumnWidth }>
+                  <Grid gutters="tiny" onClick={ () => onToggleRowVisibility && onToggleRowVisibility(data[index]) } responsive={ false } verticalAlign="middle">
+                    <GridCell>
+                      <Text textColor={ onToggleRowVisibility && hidden && 'disabled' }>{ label }</Text>
+                    </GridCell>
+                    { onToggleRowVisibility &&
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: All of the template change for inside `ChartTableLabel` should probably live in the `ChartTableLabel` component. That way the same behaviour will be available on DotPlots too.

File: packages/axiom-charts/src/BarChart/BarChartBars.js [hunk 15]
```diff
@@ -124,21 +126,22 @@ export default class BarChartBars extends Component {
...
                 </div>
-              </div>
+              </Base>
             );
           }) }
         </Bars>
 
         { benchmarkValue !== undefined && (
-          <div className="ax-bar-chart__benchmark-line-container">
+          <Base className="ax-bar-chart__benchmark-line-container">
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChartBars.js]: Is this change needed?

File: packages/axiom-charts/src/BarChart/BarChartBars.js [hunk 16]
```diff
@@ -98,7 +100,7 @@ export default class BarChartBars extends Component {
             const FinalBar = showDifferenceArea ? CombinedBar : Bar;
 
             return (
-              <div className="ax-bar-chart__bar-container" key={ color }>
+              <Base className="ax-bar-chart__bar-container" cloak={ isConcealed } key={ color }>
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChartBars.js]: How come the bar container and benchmark line are being cloak separately as opposed to cloaking this entire component at the top?

File: packages/axiom-charts/src/BarChart/BarChart.js [hunk 17]
```diff
@@ -195,45 +205,60 @@ export default class BarChart extends Component {
...
-          ) }
+          { formattedData.map(({ values, label, benchmark, hidden }, index) => {
+            const Row = hidden ? 'div' : CloakContainer;
+
+            return (<Row key={ index }>
+              <ChartTableRow hover={ index === selectedIndex }>
+                <ChartTableLabel
+                    textStrong={ index === selectedIndex }
+                    width={ labelColumnWidth }>
+                  <Grid gutters="tiny" onClick={ () => onToggleRowVisibility && onToggleRowVisibility(data[index]) } responsive={ false } verticalAlign="middle">
```
Comment (HHogg) [inline packages/axiom-charts/src/BarChart/BarChart.js]: Would it be enough just to be giving the row index here? The row data seems unnecessary and possibly harder when using this feature. See comment below in the demo component.

File: site/components/Documentation/Resources/Charts/BarChart.js [hunk 18]
```diff
@@ -8,6 +8,11 @@ import {
...
+  changedRow.hidden = !changedRow.hidden;
```
Comment (HHogg) [inline site/components/Documentation/Resources/Charts/BarChart.js]: Yeah this doesn't look right, we shouldn't be mutating the data fixture. This would be better storing this in state, similar to how it would be done inside an application (local or redux) like ... 

```jsx
const { hiddenRows } = this.state;

const handleToggleRowVisibility = (i) => 
  this.setState({ 
    hiddenRows: {
      ...hiddenRows,
      [i]: !hiddenRows[i],
    },    
  });
 
<BarChart
  data={ data.map((_, i) => ({ ..._, hidden: hiddenRows[i] })) }
  onToggleRowVisibility={ (a, b, i) => handleToggleRowVisibility(i) } />
```
